src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -253,6 +256,42 @@ private static string GetTestAgentExePath(bool requires32Bit)
             return Path.Combine(engineDir, agentName);
         }
 
+        private static void OnAgentExit(object sender, EventArgs e)
+        {
+            var process = sender as Process;
+            if (process == null)
+                return;
+
+            string errorMsg;
+
+            switch (process.ExitCode)
+            {
+                case AgentExitCodes.OK:
+                    return;
+                case AgentExitCodes.PARENT_PROCESS_TERMINATED:
+                    errorMsg = "Remote test agent believes agency process has exited.";
+                    break;
+                case AgentExitCodes.UNEXPECTED_EXCEPTION:
+                    errorMsg = "Unhandled exception on remote test agent. " +
+                               "To debug, try running with the --inprocess flag, or using --trace=debug to output logs.";
+                    break;
+                case AgentExitCodes.FAILED_TO_START_AGENT:
+                    errorMsg = "Failed to start remote test agent.";
+                    break;
+                case AgentExitCodes.NO_DEBUGGER_SECURITY:
+                    errorMsg = "Debugger could not be started on remote agent due to System.Security.Permissions.UIPermission not being set.";
+                    break;
+                case AgentExitCodes.NO_DEBUGGER_NOT_IMPLEMENTED:
+                    errorMsg = "Debugger could not be started on remote agent as not available on platform.";
+                    break;
```

Comment:
This brings to mind that the same errors can occur when the user passes the `--debug` option to the console runner. What happens in that case? What happens in general if there is an exception in the engine? We may not want to deal with it in the same PR, of course.

Reply:
`--debug` launches the debugger in the framework - I PR'd that side a few weeks back.

> What happens in general if there is an exception in the engine?

Not sure what you mean by this?